lib/plugins/esbuild/index.js

```diff
@@ -314,6 +314,22 @@ class Esbuild {
 
   async _buildProperties() {
     const defaultConfig = { bundle: true, minify: false, sourcemap: true }
+
+    const packageJson = await this._readPackageJson()
+
+    // If the user explicitly set the type to "module" then we need to set the output format to ESM
+    if (packageJson.type === 'module') {
+      defaultConfig.format = `esm`
+
+      // We will also need to make sure we include the package.json in the build
+      // for AWS Lambda to know that it is an ESM module
+      this.serverless.service.package = this.serverless.service.package || {}
+      this.serverless.service.package.patterns = [
+        ...(this.serverless.service.package.patterns || []),
+        'package.json',
```

Comment:
We shouldn't be doing it this way.

We already prepare a package.json to include in the build [here](https://github.com/serverless/serverless/blob/main/lib/plugins/esbuild/index.js#L858), we should instead make sure this is setup and included correctly.

Reply:
There was nothing wrong in the `preparePackageJson` function. It gets added to the build directory as expected, but it just wasn't being added to the zip file. I updated the PR with a fix in the zip function.